src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/RemappingCSharpCodeActionResolver.cs

```diff
@@ -0,0 +1,142 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    /// <summary>
+    /// Resolves and remaps the code action, without running formatting passes.
+    /// </summary>
+    internal class RemappingCSharpCodeActionResolver : CSharpCodeActionResolver
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly DocumentVersionCache _documentVersionCache;
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public RemappingCSharpCodeActionResolver(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver,
+            ClientNotifierServiceBase languageServer,
+            DocumentVersionCache documentVersionCache,
+            RazorDocumentMappingService documentMappingService)
+            : base(languageServer)
+        {
+            _foregroundDispatcher = foregroundDispatcher ?? throw new ArgumentNullException(nameof(foregroundDispatcher));
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+            _documentVersionCache = documentVersionCache ?? throw new ArgumentNullException(nameof(documentVersionCache));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.Remap;
+
+        public async override Task<CodeAction> ResolveAsync(
+            CSharpCodeActionParams csharpParams,
+            CodeAction codeAction,
+            CancellationToken cancellationToken)
+        {
+            if (csharpParams is null)
+            {
+                throw new ArgumentNullException(nameof(csharpParams));
+            }
+
+            if (codeAction is null)
+            {
+                throw new ArgumentNullException(nameof(codeAction));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var resolvedCodeAction = await ResolveCodeActionWithServerAsync(codeAction, cancellationToken).ConfigureAwait(false);
+            if (resolvedCodeAction.Edit?.DocumentChanges is null)
+            {
+                // Unable to resolve code action with server, return original code action
+                return codeAction;
+            }
+
+            if (resolvedCodeAction.Edit.DocumentChanges.Count() != 1)
+            {
+                // We don't yet support multi-document code actions, return original code action
```

Comment:
Hmm, should we be doing anything here in this scenario? Aka debug.assert to understand when it happens? Or strip the edit so it doesn't get applied?

Reply:
Added a Debug.Fail. 

In this case we return the *original* code action which has a null edit, not the resolved code action with the multi-document edit. Hence, no stripping of the edit is required. 😄